File: projects/04/clojure/src/factorial.clj
;;;; factorial.clj uses a driver to print out factorial of a number
;;;;
;;;; Input: n, an integer, and i, an integer that keeps track of current state in recursion
;;;; Output: the factorial of n
;;;;
;;;; Usage: clojure -m factorial
;;;;

;;;; Completed by: Justin Voss
;;;; Date: 2/4/22
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns factorial)                    ; name the program


;;;; Recieve: n, an integer
;;;; Precondition: n is an integer, and i, an integer that keeps track of current state in recursion
;;;; Return: the factorial of n, printed out
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;



;; gets passed a current state, i, and an integer n
(defn factorial [n i answer]
    (if (<= i n)              ;; checks if start is less than or equal to stop
        (do                   ;; do statement for multiple statements
        
                              ;; does recursive call of the factorial function
        (recur n (+ i 1) (* answer i))
        )
                              ;; if the current state is larger than the inputted number, it has succesfully computed the factorial.
    (if (>= i n)
      (print n "! = " answer "\n")
    )
  )
)
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Function -main() test-drives our factorial() function.
;;; Input: n, an integer
;;; Output: nothing (the output is printed in the factorial statement)
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn -main []
 (print "\nTo compute n!, enter n:") (flush)
 (let
    [ n (int(read)) ]                ; read n (an integer)
    (factorial n 2 1)
 )
)

File: projects/04/factorial.rb
#! /usr/bin/ruby
# factorial.rb computes the factorial of a number
#

# Completed by: Justin Voss, 2/4/22
#
# Input:  an integer, n
# Precondition: the number n is an integer
# Output: The factorial of a number, printed out to the console


# Computes factorial
 def factorial(n)
    answer = 1;
    i = 2;
    # for loop
    for  i in 2..n do
       answer = i * answer 
    end
    return answer.to_s
end


if __FILE__ == $0
    print "To compute n!, enter n: "
    n = gets.to_f # gets input value
    puts " #{ n } ! =  #{factorial(n)}" # prints out the statement
end